File: frontend/src/api/posts.ts
import { api } from './client'
import type { Post } from '@/types'

// 后端字段（snake_case）→ 前端类型（camelCase）
function adapt(raw: any): Post {
  return {
    slug:        raw.slug,
    title:       raw.title,
    summary:     raw.summary,
    content:     raw.content,
    tags:        raw.tags ? raw.tags.split(',').map((t: string) => t.trim()).filter(Boolean) : [],
    createdAt:   raw.created_at,
    readingTime: raw.reading_time,
    viewCount:   raw.view_count ?? 0,
    cover:       raw.cover,
  }
}

export const postsApi = {
  list(params?: { page?: number; tag?: string }): Promise<Post[]> {
    const q = new URLSearchParams()
    if (params?.page) q.set('page', String(params.page))
    if (params?.tag)  q.set('tag', params.tag)
    return api.get<any[]>(`/api/posts?${q}`).then(r => r.map(adapt))
  },

  get(slug: string): Promise<Post> {
    return api.get<any>(`/api/posts/${slug}`).then(adapt)
  },

  create(data: {
    slug: string; title: string; summary: string; content: string
    tags: string; reading_time: number; published: boolean
  }): Promise<Post> {
    return api.post<any>('/api/posts', data).then(adapt)
  },

  update(slug: string, data: Partial<{
    title: string; summary: string; content: string
    tags: string; reading_time: number; published: boolean
  }>): Promise<Post> {
    return api.patch<any>(`/api/posts/${slug}`, data).then(adapt)
  },

  delete(slug: string): Promise<void> {
    return api.delete(`/api/posts/${slug}`)
  },

  adjacent(slug: string): Promise<{ prev: { slug: string; title: string } | null; next: { slug: string; title: string } | null }> {
    return api.get(`/api/posts/${slug}/adjacent`)
  },
}

File: backend/routers/posts.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlmodel import select
from sqlalchemy.ext.asyncio import AsyncSession
from database import get_session
from models import Post, PostRead, PostCreate, PostUpdate
from auth import get_current_admin

router = APIRouter(prefix="/api/posts", tags=["文章"])


@router.get("", response_model=list[PostRead])
async def list_posts(
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=50),
    tag: str | None = None,
    session: AsyncSession = Depends(get_session),
):
    """获取文章列表（公开接口）"""
    stmt = select(Post).where(Post.published == True)
    if tag:
        stmt = stmt.where(Post.tags.contains(tag))
    stmt = stmt.order_by(Post.created_at.desc()).offset((page - 1) * limit).limit(limit)
    result = await session.execute(stmt)
    return result.scalars().all()


@router.get("/{slug}", response_model=PostRead)
async def get_post(slug: str, session: AsyncSession = Depends(get_session)):
    """获取单篇文章（公开接口），同时累计阅读量"""
    result = await session.execute(select(Post).where(Post.slug == slug))
    post = result.scalar_one_or_none()
    if not post:
        raise HTTPException(status_code=404, detail="文章不存在")
    post.view_count += 1
    session.add(post)
    await session.commit()
    await session.refresh(post)
    return post


@router.post("", response_model=PostRead, dependencies=[Depends(get_current_admin)])
async def create_post(data: PostCreate, session: AsyncSession = Depends(get_session)):
    """创建文章（需要登录）"""
    existing = await session.execute(select(Post).where(Post.slug == data.slug))
    if existing.scalar_one_or_none():
        raise HTTPException(status_code=409, detail="slug 已存在")
    post = Post(**data.model_dump())
    session.add(post)
    await session.commit()
    await session.refresh(post)
    return post


@router.patch("/{slug}", response_model=PostRead, dependencies=[Depends(get_current_admin)])
async def update_post(slug: str, data: PostUpdate, session: AsyncSession = Depends(get_session)):
    """更新文章（需要登录）"""
    result = await session.execute(select(Post).where(Post.slug == slug))
    post = result.scalar_one_or_none()
    if not post:
        raise HTTPException(status_code=404, detail="文章不存在")
    for key, val in data.model_dump(exclude_none=True).items():
        setattr(post, key, val)
    from datetime import datetime
    post.updated_at = datetime.utcnow()
    session.add(post)
    await session.commit()
    await session.refresh(post)
    return post


@router.get("/{slug}/adjacent")
async def adjacent_posts(slug: str, session: AsyncSession = Depends(get_session)):
    """获取指定文章的上一篇和下一篇（按创建时间排序）"""
    result = await session.execute(select(Post).where(Post.slug == slug, Post.published == True))
    current = result.scalar_one_or_none()
    if not current:
        raise HTTPException(404, "文章不存在")

    prev_r = await session.execute(
        select(Post).where(Post.published == True, Post.created_at < current.created_at)
        .order_by(Post.created_at.desc()).limit(1)
    )
    next_r = await session.execute(
        select(Post).where(Post.published == True, Post.created_at > current.created_at)
        .order_by(Post.created_at.asc()).limit(1)
    )
    prev_post = prev_r.scalar_one_or_none()
    next_post = next_r.scalar_one_or_none()
    return {
        "prev": {"slug": prev_post.slug, "title": prev_post.title} if prev_post else None,
        "next": {"slug": next_post.slug, "title": next_post.title} if next_post else None,
    }


@router.delete("/{slug}", dependencies=[Depends(get_current_admin)])
async def delete_post(slug: str, session: AsyncSession = Depends(get_session)):
    """删除文章（需要登录）"""
    result = await session.execute(select(Post).where(Post.slug == slug))
    post = result.scalar_one_or_none()
    if not post:
        raise HTTPException(status_code=404, detail="文章不存在")
    await session.delete(post)
    await session.commit()
    return {"message": "已删除"}
